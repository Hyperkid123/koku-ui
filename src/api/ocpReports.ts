import axios from 'axios';
import { Omit } from 'react-redux';

export interface OcpDatum {
  value: number;
  units: string;
}

export interface OcpReportValue {
  capacity?: OcpDatum;
  cost?: OcpDatum;
  cluster?: string;
  cluster_alias?: string;
  count?: OcpDatum;
  date: string;
  delta_percent?: number;
  delta_value?: number;
  derived_cost?: OcpDatum;
  infrastructure_cost?: OcpDatum;
  limit?: OcpDatum;
  node?: string;
  project?: string;
  request?: OcpDatum;
  usage: OcpDatum;
}

export interface GroupByClusterData extends Omit<OcpReportData, 'clusters'> {
  service: string;
}

export interface GroupByNodeData extends Omit<OcpReportData, 'nodes'> {
  region: string;
}

export interface GroupByProjectData extends Omit<OcpReportData, 'projects'> {
  account: string;
}

export interface OcpReportData {
  clusters?: GroupByClusterData[];
  date?: string;
  delta_percent?: number;
  delta_value?: number;
  nodes?: GroupByNodeData[];
  projects?: GroupByProjectData[];
  values?: OcpReportValue[];
}

export interface OcpReportMeta {
  delta?: {
    percent: number;
    value: number;
  };
  group_by?: {
    [group: string]: string[];
  };
  order_by?: {
    [order: string]: string;
  };
  filter?: {
    [filter: string]: any;
  };
  total?: {
    capacity?: OcpDatum;
    cost: OcpDatum;
    derived_cost: OcpDatum;
    infrastructure_cost: OcpDatum;
    limit?: OcpDatum;
    request?: OcpDatum;
    usage?: OcpDatum;
  };
  count: number;
}

export interface OcpReportLinks {
  first: string;
  previous?: string;
  next?: string;
  last: string;
}

export interface OcpReport {
  meta: OcpReportMeta;
  links: OcpReportLinks;
  data: OcpReportData[];
}

export const enum OcpReportType {
  cost = 'cost',
  cpu = 'cpu',
  memory = 'memory',
  tag = 'tag',
  volume = 'volume',
}

export const OcpReportTypePaths: Record<OcpReportType, string> = {
  [OcpReportType.cost]: 'reports/openshift/costs/',
  [OcpReportType.cpu]: 'reports/openshift/compute/',
  [OcpReportType.memory]: 'reports/openshift/memory/',
  [OcpReportType.tag]: 'tags/openshift/',
  [OcpReportType.volume]: 'reports/openshift/volumes/',
};

export function runReport(reportType: OcpReportType, query: string) {
  const path = OcpReportTypePaths[reportType];
  return axios.get<OcpReport>(`${path}?${query}`);
}
