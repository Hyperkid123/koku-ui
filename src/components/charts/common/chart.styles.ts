import {
  global_Color_dark_100,
  global_Color_light_100,
  global_primary_color_100,
  global_primary_color_200,
  global_success_color_100,
  global_success_color_200,
} from '@patternfly/react-tokens';
import { css } from 'emotion';
import React from 'react';

export const chartStyles = {
  padding: 8,
  group: {
    data: { strokeWidth: 2, fillOpacity: 0.4 },
  },
  tooltipText: {
    fontSize: '14px',
    fill: global_Color_light_100.value,
  } as React.CSSProperties,
  tooltipFlyout: { fill: global_Color_dark_100.value } as React.CSSProperties,
  previousMonth: {
    data: {
      fill: global_success_color_200.value,
      stroke: global_success_color_100.value,
    },
  },
  currentMonth: {
    data: {
      fill: global_primary_color_100.value,
      stroke: global_primary_color_200.value,
    },
  },
};

export const chartOverride = css`
  :not(foo) svg {
    overflow: visible;
  }
`;
