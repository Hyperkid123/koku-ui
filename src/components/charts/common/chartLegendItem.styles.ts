import {
  global_BorderWidth_sm,
  global_Color_200,
  global_FontSize_xs,
  global_spacer_md,
  global_spacer_sm,
} from '@patternfly/react-tokens';
import React from 'react';
import { chartStyles } from './chart.styles';

export const styles = {
  legendItem: {
    display: 'flex',
    alignItems: 'center',
    fontSize: global_FontSize_xs.value,
    color: global_Color_200.var,
    marginRight: global_spacer_md.value,
  },
  color: {
    height: 10,
    width: 10,
    marginRight: global_spacer_sm.value,
    borderWidth: global_BorderWidth_sm.value,
    borderStyle: 'solid',
  },
  currentColor: {
    backgroundColor: chartStyles.currentMonth.data.fill,
    borderColor: chartStyles.currentMonth.data.stroke,
  },
  previousColor: {
    backgroundColor: chartStyles.previousMonth.data.fill,
    borderColor: chartStyles.previousMonth.data.stroke,
  },
} as { [className: string]: React.CSSProperties };
