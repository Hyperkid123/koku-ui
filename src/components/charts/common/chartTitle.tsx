import { css } from '@patternfly/react-styles';
import { global_FontSize_xs, global_spacer_sm } from '@patternfly/react-tokens';
import React from 'react';

const styles = {
  chartTitle: {
    fontSize: global_FontSize_xs.value,
    marginBottom: global_spacer_sm.value,
  },
} as { [className: string]: React.CSSProperties };

interface ChartTitleProps {
  children: React.ReactNode;
}

const ChartTitle: React.SFC<ChartTitleProps> = ({ children }) => (
  <div className={css(styles.chartTitle)}>{children}</div>
);

export { ChartTitle, ChartTitleProps };
