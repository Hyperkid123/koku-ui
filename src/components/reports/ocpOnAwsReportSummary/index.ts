export { OcpOnAwsReportSummary } from './ocpOnAwsReportSummary';
export { OcpOnAwsReportSummaryAlt } from './ocpOnAwsReportSummaryAlt';
export { OcpOnAwsReportSummaryDetails } from './ocpOnAwsReportSummaryDetails';
export { OcpOnAwsReportSummaryItem } from './ocpOnAwsReportSummaryItem';
export { OcpOnAwsReportSummaryItems } from './ocpOnAwsReportSummaryItems';
export { OcpOnAwsReportSummaryTrend } from './ocpOnAwsReportSummaryTrend';
export { OcpOnAwsReportSummaryUsage } from './ocpOnAwsReportSummaryUsage';
