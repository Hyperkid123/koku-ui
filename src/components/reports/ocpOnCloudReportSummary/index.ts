export { OcpOnCloudReportSummary } from './ocpOnCloudReportSummary';
export { OcpOnCloudReportSummaryAlt } from './ocpOnCloudReportSummaryAlt';
export {
  OcpOnCloudReportSummaryDetails,
} from './ocpOnCloudReportSummaryDetails';
export { OcpOnCloudReportSummaryItem } from './ocpOnCloudReportSummaryItem';
export { OcpOnCloudReportSummaryItems } from './ocpOnCloudReportSummaryItems';
export { OcpOnCloudReportSummaryTrend } from './ocpOnCloudReportSummaryTrend';
export { OcpOnCloudReportSummaryUsage } from './ocpOnCloudReportSummaryUsage';
