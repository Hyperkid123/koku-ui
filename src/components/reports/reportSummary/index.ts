export { ReportSummary } from './reportSummary';
export { ReportSummaryAlt } from './reportSummaryAlt';
export { ReportSummaryDetails } from './reportSummaryDetails';
export { ReportSummaryItem } from './reportSummaryItem';
export { ReportSummaryItems } from './reportSummaryItems';
export { ReportSummaryCost } from './reportSummaryCost';
export { ReportSummaryTrend } from './reportSummaryTrend';
export { ReportSummaryUsage } from './reportSummaryUsage';
