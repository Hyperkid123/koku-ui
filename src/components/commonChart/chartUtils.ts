import { AwsReport } from 'api/awsReports';
import { OcpReport } from 'api/ocpReports';
import format from 'date-fns/format';
import getDate from 'date-fns/get_date';
import startOfMonth from 'date-fns/start_of_month';
import { FormatOptions, ValueFormatter } from 'utils/formatValue';
import {
  ComputedAwsReportItem,
  getComputedAwsReportItems,
  getIdKeyForGroupBy,
} from 'utils/getComputedAwsReportItems';
import {
  ComputedOcpReportItem,
  getComputedOcpReportItems,
} from 'utils/getComputedOcpReportItems';
import { SortDirection } from 'utils/sort';

export interface ChartDatum {
  x: string | number;
  y: number;
  key: string | number;
  name?: string | number;
  units: string;
}

export const enum ChartType {
  rolling,
  daily,
  monthly,
}

export function isAwsReport(
  report: AwsReport | OcpReport
): report is AwsReport {
  const groupById = report ? getIdKeyForGroupBy(report.group_by) : 'date'; // default key
  return groupById !== 'date';
}

export function transformReport(
  report: AwsReport | OcpReport,
  type: ChartType = ChartType.daily,
  key: any = 'date'
): ChartDatum[] {
  if (!report) {
    return [];
  }
  const items = {
    report,
    idKey: key,
    sortKey: 'id',
    sortDirection: SortDirection.desc,
  } as any;
  const computedItems = isAwsReport(report)
    ? getComputedAwsReportItems(items)
    : getComputedOcpReportItems(items);
  if (type === ChartType.daily) {
    return computedItems.map(i => createDatum(i.total, i, key));
  }

  if (type === ChartType.monthly) {
    return computedItems.map(i => createDatum(i.total, i, key));
  }

  return computedItems.reduce<ChartDatum[]>((acc, d) => {
    const prevValue = acc.length ? acc[acc.length - 1].y : 0;
    return [...acc, createDatum(prevValue + d.total, d, key)];
  }, []);
}

export function createDatum(
  value: number,
  computedItem: ComputedAwsReportItem | ComputedOcpReportItem,
  idKey = 'date'
): ChartDatum {
  const xVal = idKey === 'date' ? getDate(computedItem.id) : computedItem.label;
  return {
    x: xVal,
    y: parseFloat(value.toFixed(2)),
    key: computedItem.id,
    name: computedItem.id,
    units: computedItem.units,
  };
}

export function getDatumDateRange(datums: ChartDatum[]): [Date, Date] {
  if (!datums.length) {
    const today = new Date();
    const firstOfMonth = startOfMonth(today);
    return [firstOfMonth, today];
  }

  const start = new Date(datums[0].key + 'T00:00:00');
  const end = new Date(datums[datums.length - 1].key + 'T00:00:00');
  return [start, end];
}

export function getDateRangeString(datums: ChartDatum[]) {
  const [start, end] = getDatumDateRange(datums);
  const monthName = format(start, 'MMM');
  const startDate = getDate(start);
  const endDate = getDate(end);
  return `${monthName} ${getDate(start)}${
    startDate !== endDate ? ` - ${endDate}` : ''
  }`;
}

export function getTooltipLabel(
  datum: ChartDatum,
  formatValue: ValueFormatter,
  formatOptions?: FormatOptions,
  idKey: any = 'date'
) {
  if (!datum.key) {
    return '';
  }
  if (idKey === 'date') {
    const date = format(datum.key, 'MMM D YYYY');
    return `${date}: ${formatValue(datum.y, datum.units, formatOptions)}`;
  }
  return datum.key.toString();
}
