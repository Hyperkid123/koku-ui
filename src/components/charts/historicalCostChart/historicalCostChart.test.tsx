jest.mock('date-fns/format');

import { Chart, ChartArea } from '@patternfly/react-charts';
import { OcpReport, OcpReportData } from 'api/reports/ocpReports';
import * as utils from 'components/charts/common/chartUtils';
import formatDate from 'date-fns/format';
import { shallow } from 'enzyme';
import React from 'react';
import {
  HistoricalCostChart,
  HistoricalCostChartProps,
} from './historicalCostChart';

const currentMonthReport: OcpReport = createReport('1-15-18');
const previousMonthReport: OcpReport = createReport('12-15-17');

const currentCostData = utils.transformReport(
  currentMonthReport,
  utils.ChartType.daily,
  'date',
  'cost'
);
const currentInfrastructureCostData = utils.transformReport(
  currentMonthReport,
  utils.ChartType.daily,
  'date',
  'infrastructureCost'
);
const previousCostData = utils.transformReport(
  previousMonthReport,
  utils.ChartType.daily,
  'date',
  'cost'
);
const previousInfrastructureCostData = utils.transformReport(
  previousMonthReport,
  utils.ChartType.daily,
  'date',
  'infrastructureCost'
);

jest.spyOn(utils, 'getTooltipLabel');

const getTooltipLabel = utils.getTooltipLabel as jest.Mock;

const props: HistoricalCostChartProps = {
  currentCostData,
  currentInfrastructureCostData,
  height: 100,
  formatDatumValue: jest.fn(),
  formatDatumOptions: {},
  previousCostData,
  previousInfrastructureCostData,
  title: 'Usage Title',
};

test('reports are formatted to datums', () => {
  const view = shallow(<HistoricalCostChart {...props} />);
  const charts = view.find(ChartArea);
  expect(charts.length).toBe(4);
  expect(charts.at(0).prop('data')).toMatchSnapshot('current month cost data');
  expect(charts.at(1).prop('data')).toMatchSnapshot(
    'current month infrastructure cost data'
  );
  expect(charts.at(2).prop('data')).toMatchSnapshot('previous month cost data');
  expect(charts.at(3).prop('data')).toMatchSnapshot(
    'previous month infrastructure cost data'
  );
});

test('null previous and current reports are handled', () => {
  const view = shallow(
    <HistoricalCostChart
      {...props}
      currentCostData={null}
      currentInfrastructureCostData={null}
      previousCostData={null}
      previousInfrastructureCostData={null}
    />
  );
  const charts = view.find(ChartArea);
  expect(charts.length).toBe(4);
});

test('height from props is used', () => {
  const view = shallow(<HistoricalCostChart {...props} />);
  expect(view.find(Chart).prop('height')).toBe(props.height);
});

test('labels formats with datum and value formatted from props', () => {
  const view = shallow(<HistoricalCostChart {...props} />);
  const datum: utils.ChartDatum = {
    x: 1,
    y: 1,
    key: '1-1-1',
    units: 'units',
  };
  const group = view.find(Chart);
  group.props().containerComponent.props.labels({ datum });
  expect(getTooltipLabel).toBeCalledWith(
    datum,
    expect.any(Function),
    props.formatDatumOptions,
    'date'
  );
  expect(props.formatDatumValue).toBeCalledWith(
    datum.y,
    datum.units,
    props.formatDatumOptions
  );
  expect(formatDate).toBeCalledWith(datum.key, expect.any(String));
  expect(view.find(Chart).prop('height')).toBe(props.height);
});

test('labels ignores datums without a date', () => {
  const view = shallow(<HistoricalCostChart {...props} />);
  const datum: utils.ChartDatum = {
    x: 1,
    y: 1,
    key: '',
    units: 'units',
  };
  const group = view.find(Chart);
  const value = group.props().containerComponent.props.labels({ datum });
  expect(value).toBe('');
  expect(props.formatDatumValue).not.toBeCalled();
});

test('trend is a running total', () => {
  const multiDayReport: OcpReport = {
    data: [
      createReportDataPoint('1-15-18', 1),
      createReportDataPoint('1-16-18', 2),
    ],
  };
  const view = shallow(
    <HistoricalCostChart {...props} currentCostData={multiDayReport.data} />
  );
  const charts = view.find(ChartArea);
  expect(charts.at(1).prop('data')).toMatchSnapshot('current month data');
});

test('trend is a daily value', () => {
  const multiDayReport: OcpReport = {
    data: [
      createReportDataPoint('1-15-18', 1),
      createReportDataPoint('1-16-18', 2),
    ],
  };
  const view = shallow(
    <HistoricalCostChart {...props} currentCostData={multiDayReport.data} />
  );
  const charts = view.find(ChartArea);
  expect(charts.at(1).prop('data')).toMatchSnapshot('current month data');
});

function createReport(date: string): OcpReport {
  return {
    data: [createReportDataPoint(date)],
  };
}

function createReportDataPoint(date: string, usage = 1): OcpReportData {
  return {
    date,
    values: [{ date, usage: { value: usage, units: 'unit' } }],
  };
}
