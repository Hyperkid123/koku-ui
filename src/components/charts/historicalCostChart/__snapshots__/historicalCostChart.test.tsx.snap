// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reports are formatted to datums: current month cost data 1`] = `
Array [
  Object {
    "key": "12-15-17",
    "units": "USD",
    "x": 15,
    "y": 0,
  },
]
`;

exports[`reports are formatted to datums: current month infrastructure cost data 1`] = `
Array [
  Object {
    "key": "1-15-18",
    "units": "USD",
    "x": 15,
    "y": 0,
  },
]
`;

exports[`reports are formatted to datums: previous month cost data 1`] = `
Array [
  Object {
    "key": "12-15-17",
    "units": undefined,
    "x": 15,
    "y": 0,
  },
]
`;

exports[`reports are formatted to datums: previous month infrastructure cost data 1`] = `
Array [
  Object {
    "key": "1-15-18",
    "units": undefined,
    "x": 15,
    "y": 0,
  },
]
`;

exports[`trend is a daily value: current month data 1`] = `
Array [
  Object {
    "date": "1-15-18",
    "values": Array [
      Object {
        "date": "1-15-18",
        "usage": Object {
          "units": "unit",
          "value": 1,
        },
      },
    ],
  },
  Object {
    "date": "1-16-18",
    "values": Array [
      Object {
        "date": "1-16-18",
        "usage": Object {
          "units": "unit",
          "value": 2,
        },
      },
    ],
  },
]
`;

exports[`trend is a running total: current month data 1`] = `
Array [
  Object {
    "date": "1-15-18",
    "values": Array [
      Object {
        "date": "1-15-18",
        "usage": Object {
          "units": "unit",
          "value": 1,
        },
      },
    ],
  },
  Object {
    "date": "1-16-18",
    "values": Array [
      Object {
        "date": "1-16-18",
        "usage": Object {
          "units": "unit",
          "value": 2,
        },
      },
    ],
  },
]
`;
